README.md
# THX Portal Rule

THX 可视化规则编辑器。

![](./docs/screenshot.png)

## 安装

```sh
yarn add thx-portal-rule
```

## 使用

```jsx
import { mock } from 'mockjs'
import { RuleEditor, IRuleModel } from 'thx-portal-rule'

const MOCK_MODEL_LIST: IRuleModel[] = mock({
  'list|10': [{
    'id|1000-9999': 1,
    name: '模型@cword(5)',
    code: '@word(5)',
    'fields|10': [
      {
        'id|1000-9999': 1,
        name: '字段@cword(5)',
        code: '@word(5)'
      }
    ]
  }]
}).list

export default function App () {
  const [content, setContent] = useState(MOCK_CONTENT)
  return <>
    <RuleEditor models={MOCK_MODEL_LIST} content={content}
      onChange={(nextContent) => setContent(nextContent)}
    />
    <pre>{JSON.stringify(content, null, 2)}</pre>
  </>
}
```

## 示例

```sh
yarn run examples:start
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.1-alpha.23](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.22...v0.0.1-alpha.23) (2023-07-26)

### [0.0.1-alpha.22](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.21...v0.0.1-alpha.22) (2023-04-11)


### Features

* [#4](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/issues/4) 支持更多操作符生成代码 ([f76411f](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/f76411f7c1d87f579330daf5fda8b0d4a8b680e3))
* 优化model和field和operator的联动关系逻辑 ([1c2e741](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/1c2e741fda54cd5c551d4631e419aa7a2ab91b0e))

### [0.0.1-alpha.21](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.20...v0.0.1-alpha.21) (2023-04-10)

### [0.0.1-alpha.20](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.19...v0.0.1-alpha.20) (2022-11-16)


### Features

* previewer ([cdea85d](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/cdea85de6078be1a69ed852cf3c3f9b0e1f2eb78))

### [0.0.1-alpha.19](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.18...v0.0.1-alpha.19) (2022-10-27)


### Features

* FormulaEditor ([9b20865](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/9b208657fb8de820392fe1338e18cd6d43e516ed))

### [0.0.1-alpha.18](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.17...v0.0.1-alpha.18) (2022-10-24)


### Features

* relationSelectProps ([1a8cc8b](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/1a8cc8b3b724483fd5bd2569950c39dcb3227abc))

### [0.0.1-alpha.17](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.16...v0.0.1-alpha.17) (2022-10-17)


### Features

* modelSelectProps fieldSelectProps operatorSelectProps typeSelectProps literalSetterProps ([180d219](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/180d219218cd617c856bad9068e390b1a2b309f9))

### [0.0.1-alpha.15](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.14...v0.0.1-alpha.15) (2022-09-26)


### Bug Fixes

* ts lint ([e795860](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/e795860b202b111e4b7426f058e28e84ed239dfa))

### [0.0.1-alpha.14](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.13...v0.0.1-alpha.14) (2022-09-26)

### [0.0.1-alpha.13](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/compare/v0.0.1-alpha.5...v0.0.1-alpha.13) (2022-09-26)


### Features

* 变更首个关系节点的透出逻辑 ([81efc4a](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/81efc4a5942f8b1687103cdbafa14cb7419aaabb))
* 开放model、field、operator的props透传入口 ([c52f370](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/c52f370d9e34c5e9e368afe5f908563129573381))
* 操作符下拉支持autoWidth&支持字面量样式透传 ([2d6fa21](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/2d6fa2157d96c48f76d2887469dc8566874eb2c2))
* 支持变更model清除field和operator & 去掉model的文案id ([55ecf57](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/55ecf5704489b0a1743906ee2f82a4792be2b090))
* 支持自定义默认组关系 ([f690e88](https://gitlab.alibaba-inc.com/mmfs/thx-portal-rule/commit/f690e887514ed255d20f56fe637e10fb48308f2d))

### [0.0.1-alpha.3](https://github.com/thx/thx-portal-rule/compare/v0.0.1-alpha.2...v0.0.1-alpha.3) (2022-08-04)


### Bug Fixes

* setter width ([d8c1dcf](https://github.com/thx/thx-portal-rule/commit/d8c1dcf148f12dc00a460de9134c70b8c6b4e307))

### [0.0.1-alpha.2](https://github.com/thx/thx-portal-rule/compare/v0.0.1-alpha.1...v0.0.1-alpha.2) (2022-07-28)

### [0.0.1-alpha.1](https://github.com/thx/thx-portal-rule/compare/v0.0.2-0...v0.0.1-alpha.1) (2022-07-26)

### 0.0.2-0 (2022-07-26)


src/types/index.ts
/* eslint-disable no-use-before-define */
/// <reference types="react" />

import React from 'react'

/** 规则节点类型 */
export enum IRuleNodeType {
  GROUP = 'GROUP',
  CONDITION = 'CONDITION'
}

// MO FIXED 似乎应该是 IRuleGroupNodeRelation
/** 规则分组节点之间的关系 */
export enum IRuleGroupNodeRelation {
  AND = 'AND',
  OR = 'OR'
}

/** 公式分组节点之间的关系 */
export enum IFormulaGroupNodeRelation {
  ADD = 'ADD',
  SUBTRACT = 'SUBTRACT',
  MULTIPLY = 'MULTIPLY',
  DIVIDE = 'DIVIDE'
}

// MO TODO https://www.chaijs.com/api/bdd/#method_not
// Quantifiers asterisk *
/** 操作符 */
export enum IRuleConditionOperator {
  EQUAL = 'EQUAL', // 等于
  NOT_EQUAL = 'NOT_EQUAL', // 不等于
  LESS_THAN = 'LESS_THAN', // 小于
  LESS_THAN_OR_EQUAL = 'LESS_THAN_OR_EQUAL', // 小于等于
  GREATER_THAN = 'GREATER_THAN', // 大于
  GREATER_THAN_OR_EQUAL = 'GREATER_THAN_OR_EQUAL', // 大于等于

  /** https://www.chaijs.com/api/bdd/#method_include */
  INCLUDE = 'INCLUDE', // 包含
  NOT_INCLUDE = 'NOT_INCLUDE', // 不包含

  /** https://www.chaijs.com/api/bdd/#method_within */
  WITHIN = 'WITHIN',
  NOT_WITHIN = 'NOT_WITHIN',

  BEGIN_WITH = 'BEGIN_WITH',
  NOT_BEGIN_WITH = 'NOT_BEGIN_WITH',
  END_WITH = 'END_WITH',
  NOT_END_WITH = 'NOT_END_WITH',

  /** https://www.chaijs.com/api/bdd/#method_match */
  MATCH = 'MATCH',
  NOT_MATCH = 'NOT_MATCH',

  /** Asserts that the target is a member of the given array. */
  ONE_OF = 'ONE_OF', // 成员
  NOT_ONE_OF = 'NOT_ONE_OF', // 成员
  /** Asserts that the target array has the same members as the given array. */
  MEMBERS = 'MEMBERS', // 成员
  NOT_MEMBERS = 'NOT_MEMBERS', // 成员

  /** 改变 */
  CHANGE = 'CHANGE',
  /** 不改变 */
  NOT_CHANGE = 'NOT_CHANGE',

}

/** 操作符集合 */
export interface IOperatorMap {
  [fieldType: string]: IOperatorMapItem[];
}

/** 单个操作符 */
export interface IOperatorMapItem {
  label: string;
  value: IRuleConditionOperator | any;
  code?: (node: IRuleConditionNode, left: string, right: string) => string;
}

/** 表达式右侧类型 */
export enum IRuleMode {
  LITERAL = 'LITERAL',
  MODEL = 'MODEL'
}

export enum IExpressionType {
  LITERAL = 'LITERAL',
  MODEL = 'MODEL'
}

export interface ILiteralExpression {
  type: IExpressionType.LITERAL;
  value?: string | any;
}

export interface IMemberExpression {
  type: IExpressionType.MODEL;
  modelId?: number;
  modelName?: string;
  modelCode?: string;
  fieldId?: number;
  fieldName?: string;
  fieldCode?: string;
  fieldType?: IRuleFieldType;
  value?: string | any;
}

export interface IRuleGroupNode {
  id: number;
  parentId?: number;
  type: IRuleNodeType.GROUP;
  relation?: IRuleGroupNodeRelation | IFormulaGroupNodeRelation,
  children?: (IRuleConditionNode | IRuleGroupNode)[];
}

export interface IRuleConditionNode {
  id?: number;
  parentId?: number;
  type: IRuleNodeType.CONDITION;
  left?: IMemberExpression;
  right?: ILiteralExpression | IMemberExpression;
  operator?: IRuleConditionOperator;
}

/**
 * 模型 + 字段
 */
export interface IRuleModel {
  id: number;
  name: string;
  code: string;
  fields: IRuleField[]
}

export type ISetter =
  'BoolSetter' | 'NumberSetter' | 'RangeSetter' | 'TextSetter' |
  'DateSetter' | 'DateTimeSetter' | 'YearSetter' | 'MonthSetter' | 'RangeDateSetter' | 'RangeDateTimeSetter' | 'TimeSetter' |
  React.ReactElement

export interface IRuleField {
  /** 唯一标识。TODO 似乎会导致数据转换非常麻烦。 */
  id: number;
  name: string;
  code: string;
  tooltip?: any;
  type?: IRuleFieldType;
  /** 配置项的设置器 */
  setter?: ISetter;
  /** 针对设置器的配置 */
  setterProps?: { [key: string]: any; };
}

// MO FIXED IFieldType => IRuleFieldType，需要暴露这个类型声明吗？不需要，减少顶层概念的数量
export type IRuleFieldType = '*' | 'BOOLEAN' | 'NUMBER' | 'STRING' | 'DATE' | 'DATETIME' | 'COLLECTION' | string


src/transform/rule2data.ts
import { IExpressionType, IRuleConditionNode, IRuleGroupNode, IRuleNodeType } from '../types/index'

function condition2data (node: IRuleConditionNode, result: { [key: string]: any; } = {}) {
  if (node.left?.modelCode) {
    result[node.left?.modelCode] = result[node.left?.modelCode] || {}
    if (node.left?.fieldCode) {
      if (!(node.left?.fieldCode in result[node.left?.modelCode])) {
        result[node.left?.modelCode][node.left?.fieldCode] = ''
      }
      if (node.right?.type === IExpressionType.LITERAL) {
        result[node.left?.modelCode][node.left?.fieldCode] = node.right.value
      }
    }
  }
  if (node.right?.type === IExpressionType.MODEL) {
    if (node.right?.modelCode) {
      result[node.right?.modelCode] = result[node.right?.modelCode] || {}
      if (node.right?.fieldCode) {
        if (!(node.right?.fieldCode in result[node.right?.modelCode])) {
          result[node.right?.modelCode][node.right?.fieldCode] = ''
        }
      }
    }
  }

  return result
}

function group2data (node: IRuleGroupNode, result: { [key: string]: any; } = {}) {
  if (!node) return result
  result = result || {}
  if (node.children) {
    node.children.map(child => {
      if (child.type === IRuleNodeType.GROUP) group2data(child, result)
      if (child.type === IRuleNodeType.CONDITION) condition2data(child, result)
    })
  }
  return result
}

export function rule2data (node: IRuleGroupNode | IRuleConditionNode, result: { [key: string]: any; } = {}) {
  if (node.type === IRuleNodeType.GROUP) return group2data(node, result)
  if (node.type === IRuleNodeType.CONDITION) return condition2data(node, result)
}


src/transform/rule2expression.ts
import { FORMULA_RELATION_LIST, OPERATOR_TYPE_MAP, RULE_RELATION_LIST } from '../shared/index'
import { IExpressionType, IOperatorMap, IOperatorMapItem, IRuleConditionNode, IRuleConditionOperator, IRuleGroupNode, IRuleNodeType } from '../types/index'

// MO TODO 以下内容参考自 src/pages/sha/indicators/IndicatorTable.tsx

function literal2code (value) {
  return typeof value === 'number' ? value
    : typeof value === 'boolean' ? value
      : typeof value === 'string' ? `'${value}'`
        : value
}

function condition2expression (node: IRuleConditionNode, context = '') {
  const result: string[] = []
  // 左侧 模型.字段
  const leftExpression = `${context}${node.left?.modelCode}.${node.left?.fieldCode}`
  result.push(leftExpression)
  // 中间 操作符
  const middleOperator = node.operator
    ? (
      (IRuleConditionOperator.EQUAL === node.operator && '===') ||
      (IRuleConditionOperator.NOT_EQUAL === node.operator && '!==') ||
      (IRuleConditionOperator.LESS_THAN === node.operator && '<') ||
      (IRuleConditionOperator.LESS_THAN_OR_EQUAL === node.operator && '<=') ||
      (IRuleConditionOperator.GREATER_THAN === node.operator && '>') ||
      (IRuleConditionOperator.GREATER_THAN_OR_EQUAL === node.operator && '>=') ||
      node.operator
    )
    : ''
  if (node.operator) result.push(middleOperator)
  // 右侧 模型.字段 ｜ 字面量
  const rightExpression = node.operator
    ? (
      node.right?.type === IExpressionType.LITERAL ? literal2code(node.right.value)
        : node.right?.type === IExpressionType.MODEL ? `${context}${node.right?.modelCode}.${node.right?.fieldCode}`
          : ''
    )
    : ''
  if (node.operator) result.push(rightExpression)

  // 左侧字段的类型和中间的操作符，决定了表达式的生成方式
  // MO TODO 左侧字段类型
  // MO TODO 支持更多操作符
  if (OPERATOR_TYPE_MAP[node.left.fieldType || '*']) {
    const currentOperatorMap: IOperatorMap = OPERATOR_TYPE_MAP // operatorMap || OPERATOR_TYPE_MAP
    const dataSource: IOperatorMapItem[] = currentOperatorMap[node.left.fieldType || '*'] || currentOperatorMap['*']
    const operatorItem: IOperatorMapItem = dataSource.find(item => item.value === node.operator)
    if (operatorItem?.code) {
      return [
        '(',
        operatorItem.code(node, leftExpression, rightExpression),
        ')'
      ].join(' ')
    }
  }

  return [
    '(',
    ...result,
    // leftExpression,
    // middleOperator,
    // rightExpression,
    ')'
  ].join(' ')
}

function group2expression (node: IRuleGroupNode, context = '') {
  if (!node) return ''
  const result: any[] = []
  if (node.children) {
    node.children.map(child => {
      if (child.type === IRuleNodeType.GROUP) result.push(group2expression(child, context))
      if (child.type === IRuleNodeType.CONDITION) result.push(condition2expression(child, context))
    })
  }
  return [
    '(',
    result.join(
      ' ' +
      (
        RULE_RELATION_LIST.find(item => item.value === node.relation)?.code ||
        FORMULA_RELATION_LIST.find(item => item.value === node.relation)?.code
      ) +
      ' '
    ),
    ')'
  ].join(' ')
}

export function rule2expression (node: IRuleGroupNode | IRuleConditionNode, context = '') {
  if (node.type === IRuleNodeType.GROUP) return group2expression(node, context)
  if (node.type === IRuleNodeType.CONDITION) return condition2expression(node, context)
}


examples/react-guide-app/src/model.ts
import { mock } from 'mockjs'
import {
  IRuleGroupNode, IRuleConditionOperator, IRuleGroupNodeRelation,
  IRuleModel,
  OPERATOR_TYPE_MAP,
  IRuleNodeType,
  IExpressionType,
  ISetter,
  IRuleConditionNode,
  IOperatorMap,
  IFormulaGroupNodeRelation,
  IRuleFieldType
} from 'thx-portal-rule'

// 概念
// Program Expression left right operator
// Rule / condition

let __uuid__ = 0
export function uuid () {
  return ++__uuid__
}

const SETTER_NAME_LIST: ISetter[] = ['BoolSetter', 'NumberSetter', 'RangeSetter', 'TextSetter', 'DateSetter', 'DateTimeSetter', 'YearSetter', 'MonthSetter', 'RangeDateSetter', 'RangeDateTimeSetter', 'TimeSetter']
const SETTER_TYPE_LIST: IRuleFieldType[] = ['BOOLEAN', 'NUMBER', 'NUMBER', 'STRING', 'STRING', 'STRING', 'STRING', 'STRING', 'COLLECTION', 'COLLECTION', 'STRING']

export const MOCK_MODEL_LIST: IRuleModel[] = mock({
  'list|10': [{
    'id|1000-9999': 1,
    name: '模型@cword(5)',
    code: '@word(5)',
    [`fields|${SETTER_NAME_LIST.length}`]: [
      {
        'id|1000-9999': 1,
        name: '字段@cword(5)',
        code: '@word(5)',
        'setter|+1': SETTER_NAME_LIST,
        'type|+1': SETTER_TYPE_LIST // 'BOOLEAN'
      }
    ]
  }]
}).list.map((model: IRuleModel) => ({
  ...model,
  fields: model.fields.map(field => ({
    ...field,
    name: `${field.setter} ${field.name} #${field.code}`
  }))
}))

export const MOCK_CONTENT_SIMPLE_1: IRuleGroupNode = {
  id: uuid(),
  type: IRuleNodeType.GROUP,
  relation: IRuleGroupNodeRelation.AND,
  children: [
    {
      id: uuid(),
      type: IRuleNodeType.GROUP,
      relation: IRuleGroupNodeRelation.AND,
      children: [
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[0].id,
            modelName: MOCK_MODEL_LIST[0].name,
            modelCode: MOCK_MODEL_LIST[0].code,
            fieldId: MOCK_MODEL_LIST[0].fields[0].id,
            fieldName: MOCK_MODEL_LIST[0].fields[0].name,
            fieldCode: MOCK_MODEL_LIST[0].fields[0].code,
            fieldType: 'BOOLEAN'
          },
          right: {
            type: IExpressionType.LITERAL,
            value: true
          },
          operator: OPERATOR_TYPE_MAP['*'][0].value
        }
      ]
    }
  ]
}

export const MOCK_CONTENT_SIMPLE_2: IRuleGroupNode = {
  id: uuid(),
  type: IRuleNodeType.GROUP,
  relation: IRuleGroupNodeRelation.AND,
  children: [
    {
      id: uuid(),
      type: IRuleNodeType.GROUP,
      relation: IRuleGroupNodeRelation.AND,
      children: [
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[0].id,
            modelName: MOCK_MODEL_LIST[0].name,
            modelCode: MOCK_MODEL_LIST[0].code,
            fieldId: MOCK_MODEL_LIST[0].fields[0].id,
            fieldName: MOCK_MODEL_LIST[0].fields[0].name,
            fieldCode: MOCK_MODEL_LIST[0].fields[0].code,
            fieldType: 'BOOLEAN'
          },
          right: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[1].id,
            modelName: MOCK_MODEL_LIST[1].name,
            modelCode: MOCK_MODEL_LIST[1].code,
            fieldId: MOCK_MODEL_LIST[1].fields[1].id,
            fieldName: MOCK_MODEL_LIST[1].fields[1].name,
            fieldCode: MOCK_MODEL_LIST[1].fields[1].code
          },
          operator: OPERATOR_TYPE_MAP['*'][0].value
        }
      ]
    }
  ]
}

export const MOCK_CONTENT_SIMPLE_3: IRuleGroupNode = {
  id: uuid(),
  type: IRuleNodeType.GROUP,
  relation: IRuleGroupNodeRelation.AND,
  children: [
    {
      id: uuid(),
      type: IRuleNodeType.GROUP,
      relation: IRuleGroupNodeRelation.AND,
      children: [
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[0].id,
            modelName: MOCK_MODEL_LIST[0].name,
            modelCode: MOCK_MODEL_LIST[0].code,
            fieldId: MOCK_MODEL_LIST[0].fields[0].id,
            fieldName: MOCK_MODEL_LIST[0].fields[0].name,
            fieldCode: MOCK_MODEL_LIST[0].fields[0].code,
            fieldType: 'BOOLEAN'
          },
          right: {
            type: IExpressionType.LITERAL,
            value: true
          },
          operator: OPERATOR_TYPE_MAP['*'][0].value
        },
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[2].id,
            modelName: MOCK_MODEL_LIST[2].name,
            modelCode: MOCK_MODEL_LIST[2].code,
            fieldId: MOCK_MODEL_LIST[2].fields[2].id,
            fieldName: MOCK_MODEL_LIST[2].fields[2].name,
            fieldCode: MOCK_MODEL_LIST[2].fields[2].code,
            fieldType: 'BOOLEAN'
          },
          right: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[3].id,
            modelName: MOCK_MODEL_LIST[3].name,
            modelCode: MOCK_MODEL_LIST[3].code,
            fieldId: MOCK_MODEL_LIST[3].fields[3].id,
            fieldName: MOCK_MODEL_LIST[3].fields[3].name,
            fieldCode: MOCK_MODEL_LIST[3].fields[3].code
          },
          operator: OPERATOR_TYPE_MAP['*'][1].value
        }
      ]
    }
  ]
}

export const MOCK_CONTENT: IRuleGroupNode = {
  id: uuid(),
  type: IRuleNodeType.GROUP,
  relation: IRuleGroupNodeRelation.AND,
  children: [
    {
      id: uuid(),
      type: IRuleNodeType.GROUP,
      relation: IRuleGroupNodeRelation.AND,
      children: [
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[0].id,
            modelName: MOCK_MODEL_LIST[0].name,
            modelCode: MOCK_MODEL_LIST[0].code,
            fieldId: MOCK_MODEL_LIST[0].fields[0].id,
            fieldName: MOCK_MODEL_LIST[0].fields[0].name,
            fieldCode: MOCK_MODEL_LIST[0].fields[0].code,
            fieldType: 'BOOLEAN'
          },
          right: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[1].id,
            modelName: MOCK_MODEL_LIST[1].name,
            modelCode: MOCK_MODEL_LIST[1].code,
            fieldId: MOCK_MODEL_LIST[1].fields[1].id,
            fieldName: MOCK_MODEL_LIST[1].fields[1].name,
            fieldCode: MOCK_MODEL_LIST[1].fields[1].code
          },
          operator: OPERATOR_TYPE_MAP['*'][0].value
        },
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[2].id,
            modelName: MOCK_MODEL_LIST[2].name,
            modelCode: MOCK_MODEL_LIST[2].code,
            fieldId: MOCK_MODEL_LIST[2].fields[2].id,
            fieldName: MOCK_MODEL_LIST[2].fields[2].name,
            fieldCode: MOCK_MODEL_LIST[2].fields[2].code,
            fieldType: 'BOOLEAN'
          },
          right: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[3].id,
            modelName: MOCK_MODEL_LIST[3].name,
            modelCode: MOCK_MODEL_LIST[3].code,
            fieldId: MOCK_MODEL_LIST[3].fields[3].id,
            fieldName: MOCK_MODEL_LIST[3].fields[3].name,
            fieldCode: MOCK_MODEL_LIST[3].fields[3].code
          },
          operator: OPERATOR_TYPE_MAP['*'][1].value
        }
      ]
    },
    ...SETTER_NAME_LIST.map<IRuleConditionNode>(setter => ({
      id: uuid(),
      type: IRuleNodeType.CONDITION,
      left: {
        type: IExpressionType.MODEL,
        modelId: MOCK_MODEL_LIST[4].id,
        modelName: MOCK_MODEL_LIST[4].name,
        modelCode: MOCK_MODEL_LIST[4].code,
        fieldId: MOCK_MODEL_LIST[4].fields.find(field => field.setter === setter)?.id,
        fieldName: MOCK_MODEL_LIST[4].fields.find(field => field.setter === setter)?.name,
        fieldCode: MOCK_MODEL_LIST[4].fields.find(field => field.setter === setter)?.code
      },
      right: {
        type: IExpressionType.LITERAL,
        value: ''
      },
      operator: IRuleConditionOperator.EQUAL
    }))
  ]
}

export const MOCK_FORMULA_SIMPLE_1: IRuleGroupNode = {
  id: uuid(),
  type: IRuleNodeType.GROUP,
  relation: IFormulaGroupNodeRelation.ADD,
  children: [
    {
      id: uuid(),
      type: IRuleNodeType.CONDITION,
      left: {
        type: IExpressionType.MODEL,
        modelId: MOCK_MODEL_LIST[0].id,
        modelName: MOCK_MODEL_LIST[0].name,
        modelCode: MOCK_MODEL_LIST[0].code,
        fieldId: MOCK_MODEL_LIST[0].fields[0].id,
        fieldName: MOCK_MODEL_LIST[0].fields[0].name,
        fieldCode: MOCK_MODEL_LIST[0].fields[0].code,
        fieldType: 'BOOLEAN'
      }
    },
    {
      id: uuid(),
      type: IRuleNodeType.CONDITION,
      left: {
        type: IExpressionType.MODEL,
        modelId: MOCK_MODEL_LIST[1].id,
        modelName: MOCK_MODEL_LIST[1].name,
        modelCode: MOCK_MODEL_LIST[1].code,
        fieldId: MOCK_MODEL_LIST[1].fields[1].id,
        fieldName: MOCK_MODEL_LIST[1].fields[1].name,
        fieldCode: MOCK_MODEL_LIST[1].fields[1].code,
        fieldType: 'BOOLEAN'
      }
    }
  ]
}

export const MOCK_FORMULA: IRuleGroupNode = {
  id: uuid(),
  type: IRuleNodeType.GROUP,
  relation: IFormulaGroupNodeRelation.ADD,
  children: [
    {
      id: uuid(),
      type: IRuleNodeType.GROUP,
      relation: IFormulaGroupNodeRelation.ADD,
      children: [
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[0].id,
            modelName: MOCK_MODEL_LIST[0].name,
            modelCode: MOCK_MODEL_LIST[0].code,
            fieldId: MOCK_MODEL_LIST[0].fields[0].id,
            fieldName: MOCK_MODEL_LIST[0].fields[0].name,
            fieldCode: MOCK_MODEL_LIST[0].fields[0].code,
            fieldType: 'BOOLEAN'
          }
        },
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[1].id,
            modelName: MOCK_MODEL_LIST[1].name,
            modelCode: MOCK_MODEL_LIST[1].code,
            fieldId: MOCK_MODEL_LIST[1].fields[1].id,
            fieldName: MOCK_MODEL_LIST[1].fields[1].name,
            fieldCode: MOCK_MODEL_LIST[1].fields[1].code,
            fieldType: 'BOOLEAN'
          }
        }
      ]
    },
    {
      id: uuid(),
      type: IRuleNodeType.GROUP,
      relation: IFormulaGroupNodeRelation.SUBTRACT,
      children: [
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[2].id,
            modelName: MOCK_MODEL_LIST[2].name,
            modelCode: MOCK_MODEL_LIST[2].code,
            fieldId: MOCK_MODEL_LIST[2].fields[2].id,
            fieldName: MOCK_MODEL_LIST[2].fields[2].name,
            fieldCode: MOCK_MODEL_LIST[2].fields[2].code,
            fieldType: 'BOOLEAN'
          }
        },
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[3].id,
            modelName: MOCK_MODEL_LIST[3].name,
            modelCode: MOCK_MODEL_LIST[3].code,
            fieldId: MOCK_MODEL_LIST[3].fields[3].id,
            fieldName: MOCK_MODEL_LIST[3].fields[3].name,
            fieldCode: MOCK_MODEL_LIST[3].fields[3].code,
            fieldType: 'BOOLEAN'
          }
        }
      ]
    },
    {
      id: uuid(),
      type: IRuleNodeType.GROUP,
      relation: IFormulaGroupNodeRelation.MULTIPLY,
      children: [
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[4].id,
            modelName: MOCK_MODEL_LIST[4].name,
            modelCode: MOCK_MODEL_LIST[4].code,
            fieldId: MOCK_MODEL_LIST[4].fields[4].id,
            fieldName: MOCK_MODEL_LIST[4].fields[4].name,
            fieldCode: MOCK_MODEL_LIST[4].fields[4].code,
            fieldType: 'BOOLEAN'
          }
        },
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[5].id,
            modelName: MOCK_MODEL_LIST[5].name,
            modelCode: MOCK_MODEL_LIST[5].code,
            fieldId: MOCK_MODEL_LIST[5].fields[5].id,
            fieldName: MOCK_MODEL_LIST[5].fields[5].name,
            fieldCode: MOCK_MODEL_LIST[5].fields[5].code,
            fieldType: 'BOOLEAN'
          }
        }
      ]
    },
    {
      id: uuid(),
      type: IRuleNodeType.GROUP,
      relation: IFormulaGroupNodeRelation.DIVIDE,
      children: [
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[6].id,
            modelName: MOCK_MODEL_LIST[6].name,
            modelCode: MOCK_MODEL_LIST[6].code,
            fieldId: MOCK_MODEL_LIST[6].fields[6].id,
            fieldName: MOCK_MODEL_LIST[6].fields[6].name,
            fieldCode: MOCK_MODEL_LIST[6].fields[6].code,
            fieldType: 'BOOLEAN'
          }
        },
        {
          id: uuid(),
          type: IRuleNodeType.CONDITION,
          left: {
            type: IExpressionType.MODEL,
            modelId: MOCK_MODEL_LIST[7].id,
            modelName: MOCK_MODEL_LIST[7].name,
            modelCode: MOCK_MODEL_LIST[7].code,
            fieldId: MOCK_MODEL_LIST[7].fields[7].id,
            fieldName: MOCK_MODEL_LIST[7].fields[7].name,
            fieldCode: MOCK_MODEL_LIST[7].fields[7].code,
            fieldType: 'BOOLEAN'
          }
        }
      ]
    }
  ]
}

export const MOCK_OPERATOR_MAP: IOperatorMap = {
  '*': [
    { label: '等于', value: 'EQUAL' },
    { label: '不等于.', value: 'NOT_EQUAL' }
  ],
  'BOOLEAN': [
    { label: '等于', value: 'EQUAL' },
    { label: '不等于', value: 'NOT_EQUAL' },
    { label: '严格等于', value: 'STRICT_EQUAL' }
  ]
}
